README.md
### WebIOPi is a fully integrated Internet of Things framework for the Raspberry Pi. ###
  * Control, debug, and use your Pi's GPIO locally or remotely, from a browser or any app.
  * WebIOPi is the perfect Swiss-knife to make connected things with the Raspberry Pi.
  * Developed and provided by Eric PTAK, french R&D engineer in networks, software, and Internet of Things.

### Features ###
  * REST API over HTTP and CoAP (draft-14) with multicast support
  * Server written in Python with zero dependency
  * Supports GPIO, Serial, I2C, SPI, 1-Wire with zero dependency
  * Supports more than [30 devices](DEVICES.md) including DAC, ADC, sensors...
  * Full Python library for the Server, GPIO, Serial, I2C, SPI and devices drivers
  * Compatible with both Python 2 and 3
  * Extensible and highly customizable
  * Login/Password protection
  * Mobile devices compatible
  * Includes debug web apps
    * GPIO Header
    * GPIO List
    * Serial Monitor
    * Devices Monitor
  * Javascript client library built on top of jQuery
  * Python client library with HTTP and CoAP support

### What for ###
  * Use webiopi.GPIO library to control GPIO in your Python scripts
  * Use devices drivers to interact with DAC, ADC, sensors and more in your Python scripts
  * Use Serial, I2C and SPI webiopi classes in your Python scripts
  * Use the GPIO Header web-app to play with GPIO and debug circuits
  * Use the Serial Monitor web-app to debug Serial devices
  * Use the Devices Monitor web-app to debug devices like DAC, ADC, sensors and more
  * Create a web-app with the Javascript library to remote controls things connected on GPIO
  * Add webiopi in your existing Python script to simply add all WebIOPi features
  * Extends webiopi with your own macro
  * Extends webiopi with scripts loading facilities in a Arduino fashion way (setup, loop, destroy)
  * Use the Python Client to make Pi-2-Pi communication through the REST API
  * Use the Python Multicast Client to control multiple Pis with a single REST call
  * Use the Java Client to use the REST API from a Java application, including Android
  * Use the REST API from any technology/language

### Where to start ###
  * [Installation](INSTALL.md)
  * [Support & Discussion](http://groups.google.com/group/webiopi)
  * [Developer's Blog (Eric/trouch)](http://trouch.com)
  * [WebRobotCam Part 1 in MagPi #9](http://issuu.com/themagpi/docs/issue_9_final?mode=window) by Eric
  * [WebRobotCam Part 2 in MagPi #10](http://issuu.com/themagpi/docs/issue10final?mode=window) by Eric
  * [Video tutorial](http://www.youtube.com/watch?v=wGahWkjettw) by RaspberryPiBeginners

<p></p>

![http://trouch.com/wp-content/uploads/2012/11/webiopi-architecture.png](http://trouch.com/wp-content/uploads/2012/11/webiopi-architecture.png)

![http://trouch.com/wp-content/uploads/2012/08/webiopi-chrome.png](http://trouch.com/wp-content/uploads/2012/08/webiopi-chrome.png)

![http://trouch.com/wp-content/uploads/2013/03/webiopi-0.6-serialmon.png](http://trouch.com/wp-content/uploads/2013/03/webiopi-0.6-serialmon.png)

![http://trouch.com/wp-content/uploads/2013/03/webiopi-0.6-devices1.png](http://trouch.com/wp-content/uploads/2013/03/webiopi-0.6-devices1.png)

<p></p>

<p align='center'><a href='https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=79EM7ZV9FQ4HU'><img src='https://www.paypal.com/en_US/i/btn/btn_donateCC_LG.gif' /></a></p>

CHANGELOG.md
## 0.7.1 (02/10/15) ##
  * Added Weaved IoT Kit

## 0.7 (03/01/14) ##
  * improved touch device handling ([issue #73](https://code.google.com/p/webiopi/issues/detail?id=#73))
  * updated jQuery library to version 1.11

  * added PiFaceDigital driver
  * added PCF8591 ADC/DAC driver
  * added MCP3002 ADC driver
  * added MCP4802, MCP4812, MCP4822 ADC drivers
  * added HYT221 Humidity Sensor driver
  * added BMP180 Temperature & Pressure Sensor driver
  * added support of multiple MCP23Sxx on a single SPI CE line
  * added vref argument for MCP DAC and ADC ([issue #48](https://code.google.com/p/webiopi/issues/detail?id=#48))
  * fixed devices monitor DAC display

  * added IPv6 binding support ([issue #56](https://code.google.com/p/webiopi/issues/detail?id=#56))
  * added Content-Length fixed ([issue #83](https://code.google.com/p/webiopi/issues/detail?id=#83))
  * added HTTP and CoAP clients IP and User-Agent in logs ([issue #76](https://code.google.com/p/webiopi/issues/detail?id=#76))
  * added 'prompt' option in [HTTP](HTTP.md) section of the config file ([issue #69](https://code.google.com/p/webiopi/issues/detail?id=#69))
  * fixed macro binding with -s switch ([issue #52](https://code.google.com/p/webiopi/issues/detail?id=#52))

  * fixed Luminosity.getLux REST mapping ([issue #51](https://code.google.com/p/webiopi/issues/detail?id=#51))
  * fixed Serial monitor ([issue #49](https://code.google.com/p/webiopi/issues/detail?id=#49))
  * fixed Serial low-level/binary handling
  * fixed 1-Wire Temperature sensors error handling ([issue #68](https://code.google.com/p/webiopi/issues/detail?id=#68))

  * allows PiClients to run on others platforms than the Pi ([issue #46](https://code.google.com/p/webiopi/issues/detail?id=#46))
  * general webiopi Python library improvements

## 0.6 (03/27/13) ##
  * Added CoAP ([draft-14](http://tools.ietf.org/html/draft-ietf-core-coap-14)) implementation (Server and Client)
  * Added Python WebIOPi Client class (HTTP or CoAP with HTTP fallback)
  * Added Python WebIOPi MulticastClient class (CoAP)
  * Added Serial, I2C, SPI and 1-Wire  support
    * Python lightweight drivers with no dependency
    * Automatically load required linux modules
    * Added many device drivers, see [device support page](DEVICES.md)
  * Added [Configuration](Configuration.md) file
  * Added Serial monitor web app
  * Added Devices monitor web app
  * Added Devices abstractions Javascript class
  * Added GPIO digitalRead/Write and getFunction methods
  * Added Cache-Control REST response header
  * Added Server logging facility
  * Added jQuery-mobile
  * Improved button up/down handling on mobile devices
  * Fixed error fetching python-dev package during setup
  * Fixed webiopi-passwd missing execution flag when installing with PiStore
  * Fixed binary file serving
  * Fixed iOS6 Safari Mobile bug


## 0.5.3 (01/07/13) ##
  * Added board revision REST URI (submitted by Andreas Riegg)
  * Added encrypted passwd file to store credentials used for HTTP authentication
  * Added webiopi-passwd command-line program to generate passwd file
  * Moved demo to examples/custom folder, added examples/basic
  * Changed /dev/mem access to allow webiopi import without root privileges when not using GPIOs
  * Disabled update checker to avoid "Update available" link
  * Improved GPIO error handling
  * Fixed encoding issue with python 2.x giving a blank page
  * Fixed setup script

## 0.5.2 (12/21/12 - PiStore only) ##
  * Fixed blank page and file handling when server start at boot
  * Improved macros handling to allow zeo, one or more args
  * Added server loop helper
  * Improved setup.sh for the PiStore
  * Added play.sh for the PiStore to open the browser

## 0.5.1 (11/16/12) ##
### New Features ###
  * REV 2 boards support
  * Added setup script to ease WebIOPi install
  * Use WebIOPi in your own Python scripts
  * Login/Password protection
  * Software PWM
  * Binary sequence output

### Python Server & REST API ###
  * Usable as a library
  * Added Python 3 support
  * Removed RPi.GPIO library dependency
  * Improved security
  * Improved file serving
  * Improved REST API
  * Added ability to use custom REST macro
  * Added ability to output a single pulse
  * Added ability to output a binary sequence
  * Added software PWM

### Javascript Library ###
  * Improved and simplified
  * Added ability to create custom buttons with one or two callbacks (mousedown and mouseup)
  * Added helpers for new REST functions (macro, pulse, sequence, PWM, ...)

### Other changes ###
  * PHP Server discontinued


## 0.3 (08/27/12) ##
  * Fixed security issue in the Python server
  * Python server general improvements
  * Python server display real IP
  * Allow PHP5 < 5.4 (Works on Debian wheezy)
  * Added update tracking
  * Refactored Javascript code to be used as a library
  * Changed IDs and CSS naming
  * Added expert app


## 0.2 (08/21/12) ##
  * Initial release

python/webiopi/utils/version.py
import re
import sys

VERSION         = '0.6.5'
VERSION_STRING  = "WebIOPi/%s/Python%d.%d" % (VERSION, sys.version_info.major, sys.version_info.minor)
PYTHON_MAJOR    = sys.version_info.major
BOARD_REVISION  = 0

_MAPPING = [[], [], []]
_MAPPING[1] = ["V33", "V50", 0, "V50", 1, "GND", 4, 14, "GND", 15, 17, 18, 21, "GND", 22, 23, "V33", 24, 10, "GND", 9, 25, 11, 8, "GND", 7]
_MAPPING[2] = ["V33", "V50", 2, "V50", 3, "GND", 4, 14, "GND", 15, 17, 18, 27, "GND", 22, 23, "V33", 24, 10, "GND", 9, 25, 11, 8, "GND", 7]


try:
    with open("/proc/cpuinfo") as f:
        rc = re.compile("Revision\s*:\s(.*)\n")
        info = f.read()
        result = rc.search(info)
        if result != None:
            hex_cpurev = result.group(1)
            if hex_cpurev.startswith("1000"):
                hex_cpurev = hex_cpurev[-4:]
            cpurev = int(hex_cpurev, 16)
            BOARD_REVISION = 1 if (cpurev < 4) else 2
        
except:
    pass

MAPPING = _MAPPING[BOARD_REVISION]


python/webiopi/devices/sensor/hytXXX.py
#   Copyright 2014 Michael Burget, Eric PTAK
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

from time import sleep
from webiopi.utils.types import toint
from webiopi.devices.i2c import I2C
from webiopi.devices.sensor import Temperature,Humidity

class HYT221(I2C, Temperature, Humidity):
    VAL_RETRIES = 30
    
    def __init__(self, slave=0x28):
        I2C.__init__(self, toint(slave))
        self.__startMeasuring__()
        
    def __str__(self):
        return "HYT221(slave=0x%02X)" % self.slave
            
    def __family__(self):
        return [Temperature.__family__(self), Humidity.__family__(self)]
    
    def __startMeasuring__(self):
      self.writeByte(0x0)
      
    def readRawData(self):
        self.__startMeasuring__()
        for i in range(self.VAL_RETRIES):
            #C-code example from sensor manufacturer suggest to wait 100ms (Duration of the measurement)
            # no to get the very last measurement shoudn't be a problem -> wait 10ms
            # try a read every 10 ms for maximum VAL_RETRIES times
            sleep(.01)
            data_bytes=self.readBytes(4)
            stale_bit = (data_bytes[0] & 0b01000000) >> 6
            if (stale_bit == 0):    
                raw_t = ((data_bytes[2] << 8) | data_bytes[3]) >> 2
                raw_h = ((data_bytes[0] & 0b00111111) << 8) | data_bytes[1]
                return (raw_t, raw_h)

        #Stale was never 0, so datas are not actual
        raise Exception("HYT221(slave=0x%02X): data fetch timeout" % self.slave)
            
    
    def __getCelsius__(self):
        (raw_t, raw_h) = self.readRawData()
        if raw_t < 0x3FFF:
            return (raw_t * 165.0 / 2**14) - 40.0
        else:
            raise ValueError("Temperature value out of range (RawValue=0x%04X Max:0x3FFF)" % raw_t)
    
    def __getFahrenheit__(self):
        return self.Celsius2Fahrenheit()

    def __getKelvin__(self):
        return self.Celsius2Kelvin()

    def __getHumidity__(self):
        (raw_t, raw_h) = self.readRawData()
        if raw_h < 0x3FFF:
            return raw_h / 2**14
        else:
            raise ValueError("Humidity value out of range (RawValue=0x%04X Max:0x3FFF)" % raw_h)
       


python/webiopi/devices/analog/mcp3x0x.py
#   Copyright 2012-2013 Eric Ptak - trouch.com
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

from webiopi.utils.types import toint
from webiopi.devices.spi import SPI
from webiopi.devices.analog import ADC

class MCP3X0X(SPI, ADC):
    def __init__(self, chip, channelCount, resolution, name):
        SPI.__init__(self, toint(chip), 0, 8, 10000)
        ADC.__init__(self, channelCount, resolution, 3.3)
        self.name = name
        self.MSB_MASK = 2**(resolution-8) - 1

    def __str__(self):
        return "%s(chip=%d)" % (self.name, self.chip)

    def __analogRead__(self, channel, diff):
        data = self.__command__(channel, diff)
        r = self.xfer(data)
        return ((r[1] & self.MSB_MASK) << 8) | r[2]
    
class MCP300X(MCP3X0X):
    def __init__(self, chip, channelCount, name):
        MCP3X0X.__init__(self, chip, channelCount, 10, name)

    def __command__(self, channel, diff):
        d = [0x00, 0x00, 0x00]
        d[0] |= 1
        d[1] |= (not diff) << 7
        d[1] |= ((channel >> 2) & 0x01) << 6
        d[1] |= ((channel >> 1) & 0x01) << 5
        d[1] |= ((channel >> 0) & 0x01) << 4
        return d
        
class MCP3002(MCP300X):
    def __init__(self, chip=0):
        MCP300X.__init__(self, chip, 2, "MCP3002")

    def __analogRead__(self, channel, diff):
        data = self.__command__(channel, diff)
        r = self.xfer(data)

        # Format of return is
        # 1 empty bit
        # 1 null bit
        # 10 ADC bits
        return (r[0]<<14 | r[1]<<6 | r[2]>>2) & ((2**self._analogResolution)-1)

    def __command__(self, channel, diff):
        d = [0x00, 0x00, 0x00]
        d[0] |= 1                                 #start bit
        d[1] |= (not diff) << 7                   #single/differntial input
        d[1] |= channel  << 6                     #channel select
        return d

class MCP3004(MCP300X):
    def __init__(self, chip=0):
        MCP300X.__init__(self, chip, 4, "MCP3004")
        
class MCP3008(MCP300X):
    def __init__(self, chip=0):
        MCP300X.__init__(self, chip, 8, "MCP3008")
        
class MCP320X(MCP3X0X):
    def __init__(self, chip, channelCount, name):
        MCP3X0X.__init__(self, chip, channelCount, 12, name)

    def __command__(self, channel, diff):
        d = [0x00, 0x00, 0x00]
        d[0] |= 1 << 2
        d[0] |= (not diff) << 1
        d[0] |= (channel >> 2) & 0x01
        d[1] |= ((channel >> 1) & 0x01) << 7
        d[1] |= ((channel >> 0) & 0x01) << 6
        return d
    
class MCP3204(MCP320X):
    def __init__(self, chip=0):
        MCP320X.__init__(self, chip, 4, "MCP3204")
        
class MCP3208(MCP320X):
    def __init__(self, chip=0):
        MCP320X.__init__(self, chip, 8, "MCP3208")
        


python/webiopi/devices/analog/__init__.py
#   Copyright 2012-2013 Eric Ptak - trouch.com
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

from webiopi.decorators.rest import request, response
from webiopi.utils.types import M_JSON

class ADC():
    def __init__(self, channelCount, resolution, vref):
        self._analogCount = channelCount
        self._analogResolution = resolution
        self._analogMax = 2**resolution - 1
        self._analogRef = vref
    
    def __family__(self):
        return "ADC"

    def checkAnalogChannel(self, channel):
        if not 0 <= channel < self._analogCount:
            raise ValueError("Channel %d out of range [%d..%d]" % (channel, 0, self._analogCount-1))

    def checkAnalogValue(self, value):
        if not 0 <= value <= self._analogMax:
            raise ValueError("Value %d out of range [%d..%d]" % (value, 0, self._analogMax))
    
    @request("GET", "analog/count")
    @response("%d")
    def analogCount(self):
        return self._analogCount

    @request("GET", "analog/resolution")
    @response("%d")
    def analogResolution(self):
        return self._analogResolution
    
    @request("GET", "analog/max")
    @response("%d")
    def analogMaximum(self):
        return int(self._analogMax)
    
    @request("GET", "analog/vref")
    @response("%.2f")
    def analogReference(self):
        return self._analogRef
    
    def __analogRead__(self, channel, diff):
        raise NotImplementedError
    
    @request("GET", "analog/%(channel)d/integer")
    @response("%d")
    def analogRead(self, channel, diff=False):
        self.checkAnalogChannel(channel)
        return self.__analogRead__(channel, diff)
    
    @request("GET", "analog/%(channel)d/float")
    @response("%.2f")
    def analogReadFloat(self, channel, diff=False):
        return self.analogRead(channel, diff) / float(self._analogMax)
    
    @request("GET", "analog/%(channel)d/volt")
    @response("%.2f")
    def analogReadVolt(self, channel, diff=False):
        if self._analogRef == 0:
            raise NotImplementedError
        return self.analogReadFloat(channel, diff) * self._analogRef
    
    @request("GET", "analog/*/integer")
    @response(contentType=M_JSON)
    def analogReadAll(self):
        values = {}
        for i in range(self._analogCount):
            values[i] = self.analogRead(i)
        return values
            
    @request("GET", "analog/*/float")
    @response(contentType=M_JSON)
    def analogReadAllFloat(self):
        values = {}
        for i in range(self._analogCount):
            values[i] = float("%.2f" % self.analogReadFloat(i))
        return values
    
    @request("GET", "analog/*/volt")
    @response(contentType=M_JSON)
    def analogReadAllVolt(self):
        values = {}
        for i in range(self._analogCount):
            values[i] = float("%.2f" % self.analogReadVolt(i))
        return values
    
class DAC(ADC):
    def __init__(self, channelCount, resolution, vref):
        ADC.__init__(self, channelCount, resolution, vref)
    
    def __family__(self):
        return "DAC"
    
    def __analogWrite__(self, channel, value):
        raise NotImplementedError
    
    @request("POST", "analog/%(channel)d/integer/%(value)d")
    @response("%d")    
    def analogWrite(self, channel, value):
        self.checkAnalogChannel(channel)
        self.checkAnalogValue(value)
        self.__analogWrite__(channel, value)
        return self.analogRead(channel)
    
    @request("POST", "analog/%(channel)d/float/%(value)f")        
    @response("%.2f")    
    def analogWriteFloat(self, channel, value):
        self.analogWrite(channel, int(value * self._analogMax))
        return self.analogReadFloat(channel)
    
    @request("POST", "analog/%(channel)d/volt/%(value)f")        
    @response("%.2f")    
    def analogWriteVolt(self, channel, value):
        self.analogWriteFloat(channel, value /self._analogRef)
        return self.analogReadVolt(channel)
    

class PWM():
    def __init__(self, channelCount, resolution, frequency):
        self._pwmCount = channelCount
        self._pwmResolution = resolution
        self._pwmMax = 2**resolution - 1
        self.frequency = frequency
        self.period = 1.0/frequency
        
        # Futaba servos standard
        self.servo_neutral = 0.00152
        self.servo_travel_time = 0.0004
        self.servo_travel_angle = 45.0
        
        self.reverse = [False for i in range(channelCount)]
         
    def __family__(self):
        return "PWM"

    def checkPWMChannel(self, channel):
        if not 0 <= channel < self._pwmCount:
            raise ValueError("Channel %d out of range [%d..%d]" % (channel, 0, self._pwmCount-1))

    def checkPWMValue(self, value):
        if not 0 <= value <= self._pwmMax:
            raise ValueError("Value %d out of range [%d..%d]" % (value, 0, self._pwmMax))
    
    def __pwmRead__(self, channel):
        raise NotImplementedError
    
    def __pwmWrite__(self, channel, value):
        raise NotImplementedError
    
    @request("GET", "pwm/count")
    @response("%d")
    def pwmCount(self):
        return self._pwmCount

    @request("GET", "pwm/resolution")
    @response("%d")
    def pwmResolution(self):
        return self._pwmResolution
    
    @request("GET", "pwm/max")
    @response("%d")
    def pwmMaximum(self):
        return int(self._pwmMax)
    
    @request("GET", "pwm/%(channel)d/integer")
    @response("%d")
    def pwmRead(self, channel):
        self.checkPWMChannel(channel)
        return self.__pwmRead__(channel)
    
    @request("GET", "pwm/%(channel)d/float")
    @response("%.2f")
    def pwmReadFloat(self, channel):
        return self.pwmRead(channel) / float(self._pwmMax)
    
    @request("POST", "pwm/%(channel)d/integer/%(value)d")
    @response("%d")    
    def pwmWrite(self, channel, value):
        self.checkPWMChannel(channel)
        self.checkPWMValue(value)
        self.__pwmWrite__(channel, value)
        return self.pwmRead(channel)
    
    @request("POST", "pwm/%(channel)d/float/%(value)f")        
    @response("%.2f")    
    def pwmWriteFloat(self, channel, value):
        self.pwmWrite(channel, int(value * self._pwmMax))
        return self.pwmReadFloat(channel)
    
    def getReverse(self, channel):
        self.checkChannel(channel)
        return self.reverse[channel]
    
    def setReverse(self, channel, value):
        self.checkChannel(channel)
        self.reverse[channel] = value
        return value
    
    def RatioToAngle(self, value):
        f = value
        f *= self.period
        f -= self.servo_neutral
        f *= self.servo_travel_angle
        f /= self.servo_travel_time
        return f

    def AngleToRatio(self, value):
        f = value
        f *= self.servo_travel_time
        f /= self.servo_travel_angle
        f += self.servo_neutral
        f /= self.period
        return f
    
    @request("GET", "pwm/%(channel)d/angle")
    @response("%.2f")
    def pwmReadAngle(self, channel):
        f = self.pwmReadFloat(channel)
        f = self.RatioToAngle(f)
        if self.reverse[channel]:
            f = -f
        else:
            f = f
        return f
        
    @request("POST", "pwm/%(channel)d/angle/%(value)f")
    @response("%.2f")
    def pwmWriteAngle(self, channel, value):
        if self.reverse[channel]:
            f = -value
        else:
            f = value
        f = self.AngleToRatio(f)
        self.pwmWriteFloat(channel, f)
        return self.pwmReadAngle(channel)

    @request("GET", "pwm/*")
    @response(contentType=M_JSON)
    def pwmWildcard(self):
        values = {}
        for i in range(self._pwmCount):
            val = self.pwmReadFloat(i)
            values[i] = {}
            values[i]["float"] = float("%.2f" % val)
            values[i]["angle"] = float("%.2f" % self.RatioToAngle(val))
        return values
    
DRIVERS = {}
DRIVERS["ads1x1x"] = ["ADS1014", "ADS1015", "ADS1114", "ADS1115"]
DRIVERS["mcp3x0x"] = ["MCP3002", "MCP3004", "MCP3008", "MCP3204", "MCP3208"]
DRIVERS["mcp4725"] = ["MCP4725"]
DRIVERS["mcp48XX"] = ["MCP4802", "MCP4812", "MCP4822"]
DRIVERS["mcp492X"] = ["MCP4921", "MCP4922"]
DRIVERS["pca9685"] = ["PCA9685"]


python/webiopi/devices/analog/mcp48XX.py
#   Copyright 2013 Dagda Ltd.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

from webiopi.utils.types import toint
from webiopi.devices.spi import SPI
from webiopi.devices.analog import DAC

class MCP48XX(SPI, DAC):
    def __init__(self, chip, channelCount, resolution, name):
        SPI.__init__(self, toint(chip), 0, 8, 10000000)
        DAC.__init__(self, channelCount, resolution, 2.048)
        self.name = name
        self.buffered=False
        self.gain=False
        self.shutdown=True
        self.values = [0 for i in range(channelCount)]

    def __str__(self):
        return "%s(chip=%d)" % (self.name, self.chip)

    def int2bin(self,n,count):
        return "".join([str((n >> y) & 1 ) for y in range(count-1,-1,-1)])

    def __analogRead__(self, channel, diff=False):
        if self.shutdown:
            ret = None
        else:
            ret = self.values[channel]
        return ret

    def __analogWriteShut__(self, channel):
        self.shutdown = True
        d = [0x00, 0x00]
        d[0] |= (channel & 0x01) << 7             # bit 15 = channel
        d[0] |= 0x00 << 6                         # bit 14 = ignored
        d[0] |= 0x00 << 5                         # bit 13 = gain
        d[0] |= (self.shutdown & 0x01) << 4       # bit 12 = shutdown
        d[0] |= 0x00                              # bits 8-11 = msb data
        d[1] |= 0x00                              # bits 0 - 7 = lsb data
        
        self.writeBytes(d)
        self.values[channel] = None

    def __analogWrite__(self, channel, value):
        self.shutdown=False
        d = [0x00, 0x00]
        d[0] |= (channel & 0x01) << 7                     # bit 15 = channel
        d[0] |= (self.buffered & 0x01) << 6               # bit 14 = ignored
        d[0] |= (not self.gain & 0x01) << 5               # bit 13 = gain
        d[0] |= (not self.shutdown & 0x01) << 4           # bit 12 = shutdown
        d[0] |= value >> (self._analogResolution - 4)     # bits 8-11 = msb data
        d[1] |= ((value << (12-self._analogResolution)) & 0xFF) # bits 4 - 7 = lsb data (4802) bits 2-7 (4812) bits 0-7 (4822)                              # bits 0 - 3 = ignored                       
        
        self.writeBytes(d)
        self.values[channel] = value
       
class MCP4802(MCP48XX):
    def __init__(self, chip=0):
        MCP48XX.__init__(self, chip, 2, 8, "MCP4802")

class MCP4812(MCP48XX):
    def __init__(self, chip=0):
        MCP48XX.__init__(self, chip, 2, 10, "MCP4812")

class MCP4822(MCP48XX):
    def __init__(self, chip=0):
        MCP48XX.__init__(self, chip, 2, 12, "MCP4822")

